pkg/bmc/redfish.go

```diff
@@ -0,0 +1,96 @@
+package bmc
+
+import (
+	"net"
+	"strings"
+)
+
+func init() {
+	registerFactory("redfish", newRedfishAccessDetails)
+	registerFactory("redfish+http", newRedfishAccessDetails)
+	registerFactory("redfish+https", newRedfishAccessDetails)
+}
+
+func newRedfishAccessDetails(bmcType, portNum, hostname, path string) (AccessDetails, error) {
+	// If the hostname is an ipv6 address, it needs to be in square brackets
+	// as we are forming a URL out of it.
+	addresses, err := net.LookupIP(hostname)
+	if err == nil {
+		for _, address := range addresses {
+			if address.String() == hostname && address.To4() == nil {
+				hostname = strings.Join([]string{"[",hostname,"]"}, "")
+			}
+		}
+	}
+	return &redfishAccessDetails{
+		bmcType:  bmcType,
+		portNum:  portNum,
+		hostname: hostname,
+		path: path,
+	}, nil
+}
+
+type redfishAccessDetails struct {
+	bmcType  string
+	portNum  string
+	hostname string
+	path     string
+}
+
+const redfishDefaultScheme = "https"
+
+func (a *redfishAccessDetails) Type() string {
+	return a.bmcType
+}
+
+// NeedsMAC returns true when the host is going to need a separate
+// port created rather than having it discovered.
+func (a *redfishAccessDetails) NeedsMAC() bool {
+	// libvirt-based hosts used for dev and testing require a MAC
+	// address, specified as part of the host, but we don't want the
+	// provisioner to have to know the rules about which drivers
+	// require what so we hide that detail inside this class and just
+	// let the provisioner know that "some" drivers require a MAC and
+	// it should ask.
+	return false
```

Comment:
Must be `true`, inspection lookup won't work otherwise.

Reply:
Could you please explain this ? From the ipmi.go, it returns `return a.bmcType == "libvirt"` meaning that when we use "ipmi://xxx" addresses such as dev env, it returns false. And what about libvirt/ non-libvirt based hosts ?

Reply:
I think one of the goals is to replace the current IPMI-based dev environment with one that uses Redfish. That would mean using the Redfish driver when the type is "libvirt". And that would mean this function should have logic similar to the current IPMI implementation, at least.

Unless @dtantsur's comment means that all Redfish servers always need a MAC pre-defined in order to have their boot process managed?

Reply:
To be clear, PXE booting requires MAC address to be present. We can use introspection to discover NICs (and thus MAC addresses), however, it has a chicked-and-egg problems: without both MAC addresses and IPMI address in our records, how do we know which introspection data corresponds to which node?

TL;DR Redfish requires a MAC address with and without inspection for the time being.

Reply:
@dtantsur why is that different for redfish from IPMI (which only requires a MAC for VMs)?

Reply:
Aren't we using the IPMI driver for dev env (and not Libvirt) : https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/libvirt/templates/ironic_nodes.json.j2#L15 And in that case the needsMAC returns false. So how does the dev env works if the MAC should be required ?

Reply:
Aren't we using the IPMI driver for dev env (and not Libvirt) : https://github.com/metal3-io/metal3-dev-env/blob/master/vm-setup/roles/libvirt/templates/ironic_nodes.json.j2#L15 then used here : https://github.com/metal3-io/metal3-dev-env/blob/master/lib/common.sh#L133 . And in that case the needsMAC returns false. So how does the dev env works if the MAC should be required ?

Reply:
and what if we switched from PXE to redfish-virtual-media boot interface ?

Reply:
The scripts for building the dev environment populate the MAC address, so even if the driver isn't requiring it, it's there.

Reply:
> why is that different for redfish from IPMI (which only requires a MAC for VMs)?

Because on real hardware we can detect IPMI address from within the machine and use it for matching data to nodes. It's much more complex for Redfish.

> what if we switched from PXE to redfish-virtual-media boot interface ?

Might work, needs testing.

Reply:
> What additional work would be needed in this repo to make it work with ironic configured that way?

To clarify on my comment: inspection is independent of the boot method in ironic, it always uses PXE and requires a port or an IPMI address for matching. Virtual media only helps if we don't use inspection.

Reply:
ok thank you @dtantsur and @dhellmann . This is now fixed.